File: 72.编辑距离.cpp
/*
 * @lc app=leetcode.cn id=72 lang=cpp
 *
 * [72] 编辑距离
 */
#include <vector>
#include <string>
using namespace std;

// @lc code=start
class Solution {
public:
    int minDistance(string word1, string word2) {
        int len1 = word1.length();
        int len2 = word2.length();
        if (len1 == 0 || len2 == 0) {
            return max(len1, len2);
        }
        vector<vector<int>> dp(len1 + 1, vector<int>(len2 + 1));
        for (int i = 0; i < len1 + 1; i++) {
            dp[i][0] = i;
        }
        for (int i = 0; i < len2 + 1; i++) {
            dp[0][i] = i;
        }

        for (int i = 1; i < len1 + 1; i++) {
            for (int j = 1; j < len2 + 1; j++) {
                int left = dp[i - 1][j] + 1;
                int down = dp[i][j - 1] + 1;
                int left_down = dp[i - 1][j - 1];
                if (word1[i - 1] != word2[j - 1]) {
                    left_down++;
                }
                dp[i][j] = min(left, min(down, left_down));
            }
        }
        return dp[len1][len2];
    }
};
// @lc code=end

File: 72.编辑距离.rs
/*
 * @lc app=leetcode.cn id=72 lang=rust
 *
 * [72] 编辑距离
 */

// @lc code=start
impl Solution {
    pub fn min_distance(word1: String, word2: String) -> i32 {
        let word1: Vec<char> = word1.chars().collect();
        let word2: Vec<char> = word2.chars().collect();
        let len1 = word1.len();
        let len2 = word2.len();
        if len1 == 0 || len2 == 0 {
            return max(len1 as i32, len2 as i32);
        }
        let mut dp = vec![vec![0; len2 + 1]; len1 + 1];
        for i in 0..=len1 {
            dp[i][0] = i as i32;
        }
        for i in 0..=len2 {
            dp[0][i] = i as i32;
        }
        for i in 1..=len1 {
            for j in 1..=len2 {
                let left = dp[i - 1][j] + 1;
                let down = dp[i][j - 1] + 1;
                let mut left_down = dp[i - 1][j - 1];
                if word1[i - 1] != word2[j - 1] {
                    left_down += 1;
                }
                dp[i][j] = min(left, min(down, left_down));
            }
        }
        dp[len1][len2]
    }
}
// @lc code=end

